### DISCUSSION TITLE
better error tracking

### DISCUSSION THREAD

**Discussion by jcupitt** [libvips/libvips#2003] [Ideas]
The global error log does not work well in large scale systems where many concurrent actors may be processing images and triggering errors. 

Investigate breaking down the error log somehow, perhaps by object. 

See https://github.com/lovell/sharp/issues/902

**Comment by dlong500** [comment]
Is this something that might get some attention in an upcoming release?  I know it has been back-burnered for a while and isn't a super high priority, but it would be nice to get some enhancements to the error handling system.

**Comment by jcupitt** [comment]
Sorry, probably not any time soon. I think the next planned thing is OSS-Fuzz integration.

**Comment by jcupitt** [comment]
Copy-pasting ideas from @lovell from that issue above for reference:

Keep the C API as now for compatibility. The object-level API changes to add an extra `error` parameter to all objects, like:

```C
GError *error = NULL;

if (vips_add (a, b, &out, "error", &error, NULL)) ) {
    g_error_message (&error);
    ....
}

VipsImage *image;
if (!(image = vips_image_new_from_file (filename, "error", &error, NULL)) {
    g_error_message (&error);
    ....
}
```

Now when an object reports an error, it does so via the `error` param in the `VipsObject` base class. If this object does not have an `error` param set, the message goes to the global log.

**Comment by dlong500** [comment]
I like that idea a lot.  Maybe if I can get some free time at some point I'll work on it and see if it meets with your approval.
